x/mongo/driver/auth/sasl.go

```diff
@@ -106,17 +142,36 @@ func ConductSaslConversation(ctx context.Context, conn driver.Connection, db str
 			bsoncore.AppendInt32Element(nil, "conversationId", int32(cid)),
 			bsoncore.AppendBinaryElement(nil, "payload", 0x00, payload),
 		)
-		saslContinueCmd := operation.NewCommand(doc).Database(db).Deployment(driver.SingleConnectionDeployment{conn})
+		saslContinueCmd := operation.NewCommand(doc).Database(sc.source).Deployment(driver.SingleConnectionDeployment{conn})
 
 		err = saslContinueCmd.Execute(ctx)
 		if err != nil {
-			return newError(err, mech)
+			return newError(err, sc.mechanism)
 		}
 		rdr = saslContinueCmd.Result()
 
 		err = bson.Unmarshal(rdr, &saslResp)
 		if err != nil {
-			return newAuthError("unmarshal error", err)
+			fullErr := fmt.Errorf("unmarshal error: %v", err)
+			return newError(fullErr, sc.mechanism)
 		}
 	}
 }
+
+// ConductSaslConversation runs a full SASL conversation to authenticate the given connection.
+func ConductSaslConversation(ctx context.Context, conn driver.Connection, authSource string, client SaslClient) error {
+	conversation := newSaslConversation(client, authSource)
+
+	// Manually create the first message instead of calling conversation.FirstMessage because we don't want the message
+	// to contain fields that are only required for speculative authentication (e.g. "db").
+	saslStartDoc, err := conversation.createFirstMessage(false)
```

Comment:
Can this boolean not be passed in when we call `newSaslConversation` such that the conversation itself knows whether it's used in a speculative manner and we don't need a comment telling us to avoid a particular method that seemingly does what we need it to?

Reply:
Changed `newSaslConversation` to take a `speculative` parameter in my latest commit.